### PR TITLE
Use dataclasses to define activitypub (de)serialization

### PR DESCRIPTION
**Pull request by mouse-reeve**
apparently there's stuff in python 3 besides using parens with print??

all the activitypub code becomes a set of `dataclass`es that describe the expected shape of the activity they represent. They can be given external activity json, make it into an activity object, and validate it, or take keyword arguments from a model, make it into an activity object, and serialize it into json.

models that are mapped into activitypub use the `ActivitypubMixin` to get serialization functionality, and must define which `dataclass` they use, and how to map the model fields to the activitypub dataclass fields.

this should streamline translation from activitypub json to models and vice versa, make a sensible space for activitypub validation, and simplify model serialization.

---

My first goal right now (august 2020) is to finish this dang PR.
What's remaining:

Activity objects
- [ ] Books
- [x] Authors
- [x] Boosts
- [x] Favorites
- [x] Tagging (#141)

Collections
- [ ] Editions list
- [x] following/followers lists

Activity verbs
- [x] Create (mostly for posting statuses)
- [x] Update
- [x] Add/remove (for shelving)
- [x] Follow/unfollow
- [x] Undo (currently only used for unfavorite)

I think this'll just be a squash commit because yikes.

### CODE REVIEW

**Comment by mouse-reeve** [comment]
this has become _such_ a stream of consciousness commit log I think I'll re-commit the changes once I'm done in a more coherent way.

**Comment by cthulahoops** [comment]
This is cool, btw. I haven't really used dataclasses, and it looks like it's going to really make things cleaner.

**Comment by mouse-reeve** [comment]
My first goal right now is to finish this dang PR.
Serializers remaining:
- [ ] Books
- [ ] Authors
- [ ] Editions list
- [ ] Create
- [ ] following/followers lists
- [ ] Shelve actions
- [ ] Follow actions
- [ ] Boosts
- [ ] Favorites
- [ ] Tagging

I think this'll just be a squash commit because yikes.

**Comment by mouse-reeve** [comment]
Current status of this PR: it's not totally ready to merge, but it'd be great if anyone who is interested in looking at it and giving feedback did.

It breaks the unit tests for Signatures (sorry @cthulahoops !) which I'm still thinking about how to fix.
It also breaks the assumptions that the fedireads data connector makes about how the incoming book data will be structured.

Overall, it's got some kludginess that I'd like to iron out, but I've been spinning my wheels on it for a while.

File: fedireads/incoming.py [hunk 1]
```diff
@@ -129,21 +130,22 @@ def handle_follow(activity):
         # Duplicate follow request. Not sure what the correct behaviour is, but
         # just dropping it works for now. We should perhaps generate the
         # Accept, but then do we need to match the activity id?
-        return
+        return True
```
Comment (cthulahoops) [inline fedireads/incoming.py:133]: I'm not a fan of this pattern of returning True/False to communicate success on the handle methods. Can we use return nothing on success and use exceptions to communicate failure?

File: fedireads/incoming.py [hunk 2]
```diff
@@ -118,7 +119,7 @@ def handle_follow(activity):
     # figure out who the actor is
     user = get_or_create_remote_user(activity['actor'])
     try:
-        request = models.UserFollowRequest.objects.create(
+        relationship, _ = models.UserFollowRequest.objects.get_or_create(
```
Comment (cthulahoops) [inline fedireads/incoming.py:122]: I don't know if the get_or_create is correct here. (If it is, the code and comment below handling the duplicate needs to change.)

**Review by cthulahoops — COMMENTED** [review]
There's a lot here, I'm not really sure how to review it. So, I've just noted a couple of things jumped out at me while looking through it.

I'm quite sure why my signature test is causing issues. I see you've replaced my mock database object with a real user object.

I'd think it'd be best to get something merged, and sort out the details in further PRs. Let me know if there's anything I can do to help with this. If you want input on anything in particular, or to talk through bits of it or whatever?

File: fedireads/incoming.py (hunk 1 above)
Comment (mouse-reeve) [reply fedireads/incoming.py:133]: yeah, I think that's a good idea. I was addressing the linter wanting the methods to have return statements that are all the same type but I think raising vs returning none is a better solution

File: fedireads/incoming.py (hunk 2 above)
Comment (mouse-reeve) [reply fedireads/incoming.py:122]: you're right! this should stay as `create`

**Pull request merged**